File: testai-agent/impact/change_detector.py
"""
TestAI Agent - Change Detector

Detects and categorizes code changes from git diffs
or file comparisons for impact analysis.
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Dict, Any, Optional, Set
import re


class ChangeType(Enum):
    """Types of code changes."""
    ADDED = "added"
    MODIFIED = "modified"
    DELETED = "deleted"
    RENAMED = "renamed"
    MOVED = "moved"


@dataclass
class CodeChange:
    """A single code change."""
    file_path: str
    change_type: ChangeType
    old_path: Optional[str] = None  # For renames
    added_lines: int = 0
    deleted_lines: int = 0
    modified_functions: List[str] = field(default_factory=list)
    modified_classes: List[str] = field(default_factory=list)
    affected_imports: List[str] = field(default_factory=list)
    is_test_file: bool = False
    content_preview: str = ""


@dataclass
class ChangeSet:
    """A set of changes (e.g., from a commit)."""
    id: str
    description: str
    changes: List[CodeChange]
    timestamp: datetime
    author: str = ""
    total_additions: int = 0
    total_deletions: int = 0
    files_changed: int = 0
    metadata: Dict[str, Any] = field(default_factory=dict)


class ChangeDetector:
    """
    Detects and categorizes code changes.

    Supports:
    - Git diff parsing
    - Function/class modification detection
    - Import analysis
    - Test file identification
    """

    # Patterns for extracting code elements
    FUNCTION_PATTERNS = {
        "python": re.compile(r"^\+?\s*def\s+(\w+)\s*\(", re.MULTILINE),
        "javascript": re.compile(r"^\+?\s*(function\s+(\w+)|(\w+)\s*=\s*(async\s+)?function|\b(\w+)\s*\([^)]*\)\s*{)", re.MULTILINE),
        "typescript": re.compile(r"^\+?\s*(function\s+(\w+)|(\w+)\s*=\s*(async\s+)?function|\b(\w+)\s*\([^)]*\)\s*:)", re.MULTILINE),
    }

    CLASS_PATTERNS = {
        "python": re.compile(r"^\+?\s*class\s+(\w+)", re.MULTILINE),
        "javascript": re.compile(r"^\+?\s*class\s+(\w+)", re.MULTILINE),
        "typescript": re.compile(r"^\+?\s*class\s+(\w+)", re.MULTILINE),
    }

    IMPORT_PATTERNS = {
        "python": re.compile(r"^\+?\s*(from\s+[\w.]+\s+import|import\s+[\w.]+)", re.MULTILINE),
        "javascript": re.compile(r"^\+?\s*(import\s+.*from|require\s*\()", re.MULTILINE),
        "typescript": re.compile(r"^\+?\s*(import\s+.*from|require\s*\()", re.MULTILINE),
    }

    TEST_FILE_PATTERNS = [
        r"test_.*\.py$",
        r".*_test\.py$",
        r".*\.test\.(js|ts|jsx|tsx)$",
        r".*\.spec\.(js|ts|jsx|tsx)$",
        r"tests?/.*\.(py|js|ts)$",
        r"__tests__/.*\.(js|ts|jsx|tsx)$",
    ]

    def __init__(self):
        """Initialize the change detector."""
        self._test_patterns = [re.compile(p) for p in self.TEST_FILE_PATTERNS]

    def parse_git_diff(
        self,
        diff_content: str,
        change_id: str = "",
        description: str = "",
    ) -> ChangeSet:
        """Parse a git diff into a ChangeSet."""
        changes = []
        current_file = None
        current_change_type = None
        old_path = None
        added_lines = 0
        deleted_lines = 0
        diff_content_buffer = []

        for line in diff_content.split("\n"):
            # New file marker
            if line.startswith("diff --git"):
                # Save previous file if exists
                if current_file:
                    changes.append(self._create_change(
                        current_file,
                        current_change_type,
                        old_path,
                        added_lines,
                        deleted_lines,
                        "\n".join(diff_content_buffer),
                    ))

                # Reset for new file
                parts = line.split()
                if len(parts) >= 4:
                    current_file = parts[3][2:]  # Remove b/ prefix
                    old_path = parts[2][2:]  # Remove a/ prefix
                current_change_type = ChangeType.MODIFIED
                added_lines = 0
                deleted_lines = 0
                diff_content_buffer = []

            # New file indicator
            elif line.startswith("new file"):
                current_change_type = ChangeType.ADDED

            # Deleted file indicator
            elif line.startswith("deleted file"):
                current_change_type = ChangeType.DELETED

            # Renamed file indicator
            elif line.startswith("rename from"):
                current_change_type = ChangeType.RENAMED
                old_path = line.replace("rename from ", "")

            elif line.startswith("rename to") and current_change_type == ChangeType.RENAMED:
                current_file = line.replace("rename to ", "")

            # Count additions and deletions
            elif line.startswith("+") and not line.startswith("+++"):
                added_lines += 1
                diff_content_buffer.append(line)

            elif line.startswith("-") and not line.startswith("---"):
                deleted_lines += 1
                diff_content_buffer.append(line)

        # Save last file
        if current_file:
            changes.append(self._create_change(
                current_file,
                current_change_type,
                old_path,
                added_lines,
                deleted_lines,
                "\n".join(diff_content_buffer),
            ))

        total_additions = sum(c.added_lines for c in changes)
        total_deletions = sum(c.deleted_lines for c in changes)

        return ChangeSet(
            id=change_id or f"CS-{datetime.now().strftime('%Y%m%d%H%M%S')}",
            description=description,
            changes=changes,
            timestamp=datetime.now(),
            total_additions=total_additions,
            total_deletions=total_deletions,
            files_changed=len(changes),
        )

    def parse_file_list(
        self,
        files: List[Dict[str, Any]],
        change_id: str = "",
        description: str = "",
    ) -> ChangeSet:
        """Create a ChangeSet from a list of file changes."""
        changes = []

        for file_info in files:
            change = CodeChange(
                file_path=file_info.get("path", ""),
                change_type=ChangeType(file_info.get("type", "modified")),
                old_path=file_info.get("old_path"),
                added_lines=file_info.get("added", 0),
                deleted_lines=file_info.get("deleted", 0),
                modified_functions=file_info.get("functions", []),
                modified_classes=file_info.get("classes", []),
                is_test_file=self._is_test_file(file_info.get("path", "")),
            )
            changes.append(change)

        return ChangeSet(
            id=change_id or f"CS-{datetime.now().strftime('%Y%m%d%H%M%S')}",
            description=description,
            changes=changes,
            timestamp=datetime.now(),
            total_additions=sum(c.added_lines for c in changes),
            total_deletions=sum(c.deleted_lines for c in changes),
            files_changed=len(changes),
        )

    def extract_modified_elements(
        self,
        diff_content: str,
        language: str = "python",
    ) -> Dict[str, List[str]]:
        """Extract modified functions and classes from diff."""
        results = {
            "functions": [],
            "classes": [],
            "imports": [],
        }

        func_pattern = self.FUNCTION_PATTERNS.get(language)
        class_pattern = self.CLASS_PATTERNS.get(language)
        import_pattern = self.IMPORT_PATTERNS.get(language)

        if func_pattern:
            matches = func_pattern.findall(diff_content)
            # Handle different match group structures
            for match in matches:
                if isinstance(match, tuple):
                    # Get first non-empty group
                    name = next((m for m in match if m), None)
                else:
                    name = match
                if name and name not in results["functions"]:
                    results["functions"].append(name)

        if class_pattern:
            matches = class_pattern.findall(diff_content)
            for match in matches:
                if match and match not in results["classes"]:
                    results["classes"].append(match)

        if import_pattern:
            matches = import_pattern.findall(diff_content)
            for match in matches:
                if match and match not in results["imports"]:
                    results["imports"].append(match)

        return results

    def categorize_changes(
        self,
        changeset: ChangeSet,
    ) -> Dict[str, List[CodeChange]]:
        """Categorize changes by type."""
        categories = {
            "source_code": [],
            "test_code": [],
            "configuration": [],
            "documentation": [],
            "other": [],
        }

        config_extensions = {".json", ".yaml", ".yml", ".toml", ".ini", ".cfg"}
        doc_extensions = {".md", ".rst", ".txt", ".adoc"}

        for change in changeset.changes:
            path = change.file_path.lower()

            if change.is_test_file:
                categories["test_code"].append(change)
            elif any(path.endswith(ext) for ext in config_extensions):
                categories["configuration"].append(change)
            elif any(path.endswith(ext) for ext in doc_extensions):
                categories["documentation"].append(change)
            elif path.endswith((".py", ".js", ".ts", ".jsx", ".tsx", ".java", ".go")):
                categories["source_code"].append(change)
            else:
                categories["other"].append(change)

        return categories

    def get_affected_modules(
        self,
        changeset: ChangeSet,
    ) -> Set[str]:
        """Get the set of affected modules/packages."""
        modules = set()

        for change in changeset.changes:
            parts = change.file_path.split("/")
            # Get top-level directory as module
            if len(parts) > 1:
                modules.add(parts[0])
            # Also add parent directories for deeper files
            for i in range(len(parts) - 1):
                modules.add("/".join(parts[:i + 1]))

        return modules

    def calculate_change_risk(
        self,
        change: CodeChange,
    ) -> float:
        """Calculate risk score for a change (0-1)."""
        risk = 0.0

        # More lines changed = higher risk
        total_lines = change.added_lines + change.deleted_lines
        if total_lines > 100:
            risk += 0.3
        elif total_lines > 50:
            risk += 0.2
        elif total_lines > 10:
            risk += 0.1

        # Function/class modifications = higher risk
        if change.modified_functions:
            risk += 0.2 * min(len(change.modified_functions) / 5, 1.0)
        if change.modified_classes:
            risk += 0.2 * min(len(change.modified_classes) / 3, 1.0)

        # Deletions are riskier than additions
        if change.deleted_lines > change.added_lines:
            risk += 0.1

        # Test file changes have lower risk (for production)
        if change.is_test_file:
            risk *= 0.5

        return min(risk, 1.0)

    def _create_change(
        self,
        file_path: str,
        change_type: Optional[ChangeType],
        old_path: Optional[str],
        added_lines: int,
        deleted_lines: int,
        diff_content: str,
    ) -> CodeChange:
        """Create a CodeChange from parsed data."""
        language = self._detect_language(file_path)
        elements = self.extract_modified_elements(diff_content, language)

        return CodeChange(
            file_path=file_path,
            change_type=change_type or ChangeType.MODIFIED,
            old_path=old_path if old_path != file_path else None,
            added_lines=added_lines,
            deleted_lines=deleted_lines,
            modified_functions=elements["functions"],
            modified_classes=elements["classes"],
            affected_imports=elements["imports"],
            is_test_file=self._is_test_file(file_path),
            content_preview=diff_content[:500] if diff_content else "",
        )

    def _is_test_file(self, file_path: str) -> bool:
        """Check if a file is a test file."""
        return any(pattern.search(file_path) for pattern in self._test_patterns)

    def _detect_language(self, file_path: str) -> str:
        """Detect programming language from file extension."""
        extension_map = {
            ".py": "python",
            ".js": "javascript",
            ".jsx": "javascript",
            ".ts": "typescript",
            ".tsx": "typescript",
        }

        for ext, lang in extension_map.items():
            if file_path.endswith(ext):
                return lang

        return "python"  # Default

    def format_changeset(self, changeset: ChangeSet) -> str:
        """Format a ChangeSet as readable text."""
        lines = [
            "=" * 60,
            f"  CHANGESET: {changeset.id}",
            "=" * 60,
            "",
            f"  Description: {changeset.description}",
            f"  Timestamp: {changeset.timestamp.strftime('%Y-%m-%d %H:%M')}",
            f"  Author: {changeset.author or 'Unknown'}",
            "",
            f"  Files Changed: {changeset.files_changed}",
            f"  Lines Added: +{changeset.total_additions}",
            f"  Lines Deleted: -{changeset.total_deletions}",
            "",
            "-" * 60,
            "  CHANGES",
            "-" * 60,
        ]

        type_icons = {
            ChangeType.ADDED: "➕",
            ChangeType.MODIFIED: "📝",
            ChangeType.DELETED: "➖",
            ChangeType.RENAMED: "📛",
            ChangeType.MOVED: "📦",
        }

        for change in changeset.changes:
            icon = type_icons.get(change.change_type, "•")
            risk = self.calculate_change_risk(change)
            risk_indicator = "🔴" if risk > 0.6 else "🟡" if risk > 0.3 else "🟢"

            lines.append(f"\n  {icon} {change.file_path} {risk_indicator}")
            lines.append(f"     +{change.added_lines} -{change.deleted_lines}")

            if change.modified_functions:
                funcs = ", ".join(change.modified_functions[:5])
                if len(change.modified_functions) > 5:
                    funcs += f" (+{len(change.modified_functions) - 5} more)"
                lines.append(f"     Functions: {funcs}")

            if change.modified_classes:
                classes = ", ".join(change.modified_classes[:3])
                lines.append(f"     Classes: {classes}")

        lines.extend(["", "=" * 60])
        return "\n".join(lines)


def create_change_detector() -> ChangeDetector:
    """Create a change detector instance."""
    return ChangeDetector()

File: electron/lib/cortex/change-detector.js
/**
 * Yali Agent - Change Detector
 * Ported from testai-agent/impact/change_detector.py
 *
 * Detects and categorizes code changes from git diffs
 * or file comparisons for impact analysis.
 */

/**
 * Types of code changes
 */
const ChangeType = {
  ADDED: 'added',
  MODIFIED: 'modified',
  DELETED: 'deleted',
  RENAMED: 'renamed',
  MOVED: 'moved'
}

/**
 * Create a code change object
 */
function createCodeChange(filePath, changeType, options = {}) {
  return {
    filePath,
    changeType,
    oldPath: options.oldPath || null, // For renames
    addedLines: options.addedLines || 0,
    deletedLines: options.deletedLines || 0,
    modifiedFunctions: options.modifiedFunctions || [],
    modifiedClasses: options.modifiedClasses || [],
    affectedImports: options.affectedImports || [],
    isTestFile: options.isTestFile || false,
    contentPreview: options.contentPreview || ''
  }
}

/**
 * Create a change set (e.g., from a commit)
 */
function createChangeSet(id, description, changes, options = {}) {
  return {
    id,
    description,
    changes,
    timestamp: options.timestamp || new Date(),
    author: options.author || '',
    totalAdditions: changes.reduce((sum, c) => sum + c.addedLines, 0),
    totalDeletions: changes.reduce((sum, c) => sum + c.deletedLines, 0),
    filesChanged: changes.length,
    metadata: options.metadata || {}
  }
}

/**
 * Change Detector class
 */
class ChangeDetector {
  // Patterns for extracting code elements
  static FUNCTION_PATTERNS = {
    python: /^\+?\s*def\s+(\w+)\s*\(/gm,
    javascript: /^\+?\s*(function\s+(\w+)|(\w+)\s*=\s*(async\s+)?function|\b(\w+)\s*\([^)]*\)\s*{)/gm,
    typescript: /^\+?\s*(function\s+(\w+)|(\w+)\s*=\s*(async\s+)?function|\b(\w+)\s*\([^)]*\)\s*:)/gm
  }

  static CLASS_PATTERNS = {
    python: /^\+?\s*class\s+(\w+)/gm,
    javascript: /^\+?\s*class\s+(\w+)/gm,
    typescript: /^\+?\s*class\s+(\w+)/gm
  }

  static IMPORT_PATTERNS = {
    python: /^\+?\s*(from\s+[\w.]+\s+import|import\s+[\w.]+)/gm,
    javascript: /^\+?\s*(import\s+.*from|require\s*\()/gm,
    typescript: /^\+?\s*(import\s+.*from|require\s*\()/gm
  }

  static TEST_FILE_PATTERNS = [
    /test_.*\.py$/,
    /.*_test\.py$/,
    /.*\.test\.(js|ts|jsx|tsx)$/,
    /.*\.spec\.(js|ts|jsx|tsx)$/,
    /tests?\/.*\.(py|js|ts)$/,
    /__tests__\/.*\.(js|ts|jsx|tsx)$/
  ]

  constructor() {
    // Pre-compile test patterns
    this._testPatterns = ChangeDetector.TEST_FILE_PATTERNS
  }

  /**
   * Parse a git diff into a ChangeSet
   */
  parseGitDiff(diffContent, changeId = '', description = '') {
    const changes = []
    let currentFile = null
    let currentChangeType = null
    let oldPath = null
    let addedLines = 0
    let deletedLines = 0
    let diffContentBuffer = []

    for (const line of diffContent.split('\n')) {
      // New file marker
      if (line.startsWith('diff --git')) {
        // Save previous file if exists
        if (currentFile) {
          changes.push(this._createChange(
            currentFile,
            currentChangeType,
            oldPath,
            addedLines,
            deletedLines,
            diffContentBuffer.join('\n')
          ))
        }

        // Reset for new file
        const parts = line.split(' ')
        if (parts.length >= 4) {
          currentFile = parts[3].slice(2) // Remove b/ prefix
          oldPath = parts[2].slice(2)      // Remove a/ prefix
        }
        currentChangeType = ChangeType.MODIFIED
        addedLines = 0
        deletedLines = 0
        diffContentBuffer = []
      }

      // New file indicator
      else if (line.startsWith('new file')) {
        currentChangeType = ChangeType.ADDED
      }

      // Deleted file indicator
      else if (line.startsWith('deleted file')) {
        currentChangeType = ChangeType.DELETED
      }

      // Renamed file indicator
      else if (line.startsWith('rename from')) {
        currentChangeType = ChangeType.RENAMED
        oldPath = line.replace('rename from ', '')
      }

      else if (line.startsWith('rename to') && currentChangeType === ChangeType.RENAMED) {
        currentFile = line.replace('rename to ', '')
      }

      // Count additions and deletions
      else if (line.startsWith('+') && !line.startsWith('+++')) {
        addedLines++
        diffContentBuffer.push(line)
      }

      else if (line.startsWith('-') && !line.startsWith('---')) {
        deletedLines++
        diffContentBuffer.push(line)
      }
    }

    // Save last file
    if (currentFile) {
      changes.push(this._createChange(
        currentFile,
        currentChangeType,
        oldPath,
        addedLines,
        deletedLines,
        diffContentBuffer.join('\n')
      ))
    }

    return createChangeSet(
      changeId || `CS-${new Date().toISOString().replace(/[-:.TZ]/g, '').slice(0, 14)}`,
      description,
      changes
    )
  }

  /**
   * Create a ChangeSet from a list of file changes
   */
  parseFileList(files, changeId = '', description = '') {
    const changes = files.map(fileInfo => createCodeChange(
      fileInfo.path || '',
      fileInfo.type || ChangeType.MODIFIED,
      {
        oldPath: fileInfo.oldPath,
        addedLines: fileInfo.added || 0,
        deletedLines: fileInfo.deleted || 0,
        modifiedFunctions: fileInfo.functions || [],
        modifiedClasses: fileInfo.classes || [],
        isTestFile: this._isTestFile(fileInfo.path || '')
      }
    ))

    return createChangeSet(
      changeId || `CS-${new Date().toISOString().replace(/[-:.TZ]/g, '').slice(0, 14)}`,
      description,
      changes
    )
  }

  /**
   * Extract modified functions and classes from diff
   */
  extractModifiedElements(diffContent, language = 'python') {
    const results = {
      functions: [],
      classes: [],
      imports: []
    }

    const funcPattern = ChangeDetector.FUNCTION_PATTERNS[language]
    const classPattern = ChangeDetector.CLASS_PATTERNS[language]
    const importPattern = ChangeDetector.IMPORT_PATTERNS[language]

    if (funcPattern) {
      const regex = new RegExp(funcPattern.source, funcPattern.flags)
      let match
      while ((match = regex.exec(diffContent)) !== null) {
        // Get first non-empty capture group
        const name = match.slice(1).find(m => m)
        if (name && !results.functions.includes(name)) {
          results.functions.push(name)
        }
      }
    }

    if (classPattern) {
      const regex = new RegExp(classPattern.source, classPattern.flags)
      let match
      while ((match = regex.exec(diffContent)) !== null) {
        if (match[1] && !results.classes.includes(match[1])) {
          results.classes.push(match[1])
        }
      }
    }

    if (importPattern) {
      const regex = new RegExp(importPattern.source, importPattern.flags)
      let match
      while ((match = regex.exec(diffContent)) !== null) {
        if (match[1] && !results.imports.includes(match[1])) {
          results.imports.push(match[1])
        }
      }
    }

    return results
  }

  /**
   * Categorize changes by type
   */
  categorizeChanges(changeset) {
    const categories = {
      sourceCode: [],
      testCode: [],
      configuration: [],
      documentation: [],
      other: []
    }

    const configExtensions = new Set(['.json', '.yaml', '.yml', '.toml', '.ini', '.cfg'])
    const docExtensions = new Set(['.md', '.rst', '.txt', '.adoc'])
    const codeExtensions = ['.py', '.js', '.ts', '.jsx', '.tsx', '.java', '.go']

    for (const change of changeset.changes) {
      const path = change.filePath.toLowerCase()

      if (change.isTestFile) {
        categories.testCode.push(change)
      } else if ([...configExtensions].some(ext => path.endsWith(ext))) {
        categories.configuration.push(change)
      } else if ([...docExtensions].some(ext => path.endsWith(ext))) {
        categories.documentation.push(change)
      } else if (codeExtensions.some(ext => path.endsWith(ext))) {
        categories.sourceCode.push(change)
      } else {
        categories.other.push(change)
      }
    }

    return categories
  }

  /**
   * Get the set of affected modules/packages
   */
  getAffectedModules(changeset) {
    const modules = new Set()

    for (const change of changeset.changes) {
      const parts = change.filePath.split('/')
      // Get top-level directory as module
      if (parts.length > 1) {
        modules.add(parts[0])
      }
      // Also add parent directories for deeper files
      for (let i = 0; i < parts.length - 1; i++) {
        modules.add(parts.slice(0, i + 1).join('/'))
      }
    }

    return modules
  }

  /**
   * Calculate risk score for a change (0-1)
   */
  calculateChangeRisk(change) {
    let risk = 0.0

    // More lines changed = higher risk
    const totalLines = change.addedLines + change.deletedLines
    if (totalLines > 100) {
      risk += 0.3
    } else if (totalLines > 50) {
      risk += 0.2
    } else if (totalLines > 10) {
      risk += 0.1
    }

    // Function/class modifications = higher risk
    if (change.modifiedFunctions.length > 0) {
      risk += 0.2 * Math.min(change.modifiedFunctions.length / 5, 1.0)
    }
    if (change.modifiedClasses.length > 0) {
      risk += 0.2 * Math.min(change.modifiedClasses.length / 3, 1.0)
    }

    // Deletions are riskier than additions
    if (change.deletedLines > change.addedLines) {
      risk += 0.1
    }

    // Test file changes have lower risk (for production)
    if (change.isTestFile) {
      risk *= 0.5
    }

    return Math.min(risk, 1.0)
  }

  _createChange(filePath, changeType, oldPath, addedLines, deletedLines, diffContent) {
    const language = this._detectLanguage(filePath)
    const elements = this.extractModifiedElements(diffContent, language)

    return createCodeChange(filePath, changeType || ChangeType.MODIFIED, {
      oldPath: oldPath !== filePath ? oldPath : null,
      addedLines,
      deletedLines,
      modifiedFunctions: elements.functions,
      modifiedClasses: elements.classes,
      affectedImports: elements.imports,
      isTestFile: this._isTestFile(filePath),
      contentPreview: diffContent.slice(0, 500)
    })
  }

  _isTestFile(filePath) {
    return this._testPatterns.some(pattern => pattern.test(filePath))
  }

  _detectLanguage(filePath) {
    const extensionMap = {
      '.py': 'python',
      '.js': 'javascript',
      '.jsx': 'javascript',
      '.ts': 'typescript',
      '.tsx': 'typescript'
    }

    for (const [ext, lang] of Object.entries(extensionMap)) {
      if (filePath.endsWith(ext)) {
        return lang
      }
    }

    return 'python' // Default
  }

  /**
   * Format a ChangeSet as readable text
   */
  formatChangeset(changeset) {
    const lines = [
      '='.repeat(60),
      `  CHANGESET: ${changeset.id}`,
      '='.repeat(60),
      '',
      `  Description: ${changeset.description}`,
      `  Timestamp: ${changeset.timestamp.toISOString().slice(0, 16).replace('T', ' ')}`,
      `  Author: ${changeset.author || 'Unknown'}`,
      '',
      `  Files Changed: ${changeset.filesChanged}`,
      `  Lines Added: +${changeset.totalAdditions}`,
      `  Lines Deleted: -${changeset.totalDeletions}`,
      '',
      '-'.repeat(60),
      '  CHANGES',
      '-'.repeat(60)
    ]

    const typeIcons = {
      [ChangeType.ADDED]: '+',
      [ChangeType.MODIFIED]: 'M',
      [ChangeType.DELETED]: '-',
      [ChangeType.RENAMED]: 'R',
      [ChangeType.MOVED]: '>'
    }

    for (const change of changeset.changes) {
      const icon = typeIcons[change.changeType] || '*'
      const risk = this.calculateChangeRisk(change)
      const riskIndicator = risk > 0.6 ? '[HIGH]' : risk > 0.3 ? '[MED]' : '[LOW]'

      lines.push(`\n  ${icon} ${change.filePath} ${riskIndicator}`)
      lines.push(`     +${change.addedLines} -${change.deletedLines}`)

      if (change.modifiedFunctions.length > 0) {
        let funcs = change.modifiedFunctions.slice(0, 5).join(', ')
        if (change.modifiedFunctions.length > 5) {
          funcs += ` (+${change.modifiedFunctions.length - 5} more)`
        }
        lines.push(`     Functions: ${funcs}`)
      }

      if (change.modifiedClasses.length > 0) {
        const classes = change.modifiedClasses.slice(0, 3).join(', ')
        lines.push(`     Classes: ${classes}`)
      }
    }

    lines.push('')
    lines.push('='.repeat(60))
    return lines.join('\n')
  }

  /**
   * Get high-risk changes that need extra testing
   */
  getHighRiskChanges(changeset, threshold = 0.5) {
    return changeset.changes.filter(c => this.calculateChangeRisk(c) >= threshold)
  }

  /**
   * Get summary for test prioritization
   */
  getSummaryForTesting(changeset) {
    const categories = this.categorizeChanges(changeset)
    const highRisk = this.getHighRiskChanges(changeset)
    const modules = this.getAffectedModules(changeset)

    return {
      totalChanges: changeset.changes.length,
      sourceCodeChanges: categories.sourceCode.length,
      testCodeChanges: categories.testCode.length,
      highRiskChanges: highRisk.length,
      affectedModules: [...modules],
      allModifiedFunctions: [...new Set(changeset.changes.flatMap(c => c.modifiedFunctions))],
      allModifiedClasses: [...new Set(changeset.changes.flatMap(c => c.modifiedClasses))],
      recommendation: this._getTestingRecommendation(changeset, highRisk)
    }
  }

  _getTestingRecommendation(changeset, highRisk) {
    if (highRisk.length === 0) {
      return 'Low risk changes - standard regression testing sufficient'
    }
    if (highRisk.length > 5) {
      return 'High risk - run full test suite with focus on affected modules'
    }
    if (highRisk.some(c => c.modifiedClasses.length > 0)) {
      return 'Class modifications detected - prioritize integration tests'
    }
    return 'Moderate risk - run targeted tests on modified functions'
  }
}

/**
 * Helper to create a change detector
 */
function createChangeDetector() {
  return new ChangeDetector()
}

module.exports = {
  // Enums
  ChangeType,

  // Factory functions
  createCodeChange,
  createChangeSet,

  // Main class
  ChangeDetector,
  createChangeDetector
}
